File: codes/dp/coinExchange/coin_exchange.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

int amount[] = {1, 5, 10, 21, 25};
unordered_map<int, int> memo;

int dfs(int total) {
    if (total == 0) return 0;
    if (memo.count(total)) return memo[total];

    int res = __INT_MAX__;
    for (int coin : amount) {
        if (coin <= total) {
            res = min(res, dfs(total - coin) + 1);
        }
    }
    return memo[total] = res;
}

int main() {
    cout << "Enter the amount : ";
    int n;
    cin >> n;
    int ans = dfs(n);
    if (ans == __INT_MAX__)
        cout << "Impossible\n";
    else
        cout << "Minimum coins: " << ans << "\n";
    return 0;
}

File: codes/dp/coinExchange/coin.exchange.py
memo = {0 : [0, []]}
amount = [1, 5, 10, 21, 25]
def dfs(total) :
    if total in memo : 
        return memo[total]
    best = float('inf')
    best_path = []
        
    for coin in amount :
        if coin <= total :
            sub_total, sub_path = dfs(total - coin)
            if sub_total + 1 < best :
                best = sub_total + 1
                best_path = sub_path + [coin]

    memo[total] = [best, best_path]
    return memo[total] 

def main():
    bill = int(input("Enter the bill amount : "))
    count, bills = dfs(bill)
    print("Minimum coins will be used : ",count)
    print("Coins used : ",bills)

if __name__ == "__main__" :
    main()
